text/287-cli-deprecation-warnings.md

```diff
@@ -0,0 +1,190 @@
+we need to communicate between the framework (where those deprecated usages happen)
+and the CLI (where the warnings will be shown).
+The natural way to do that is through the Cloud Assembly.
+
+We already have a facility for adding messages to the construct tree's metadata
+(through the Node class in the `constructs` library),
+which are then [collected into the Stack Cloud Assembly artifact type](https://github.com/aws/aws-cdk/blob/34a921b9667402b6d90731f1fd9e3de1ef27f8bf/packages/%40aws-cdk/core/lib/stack-synthesizers/_shared.ts#L63-L101),
+which the [CLI later renders](https://github.com/aws/aws-cdk/blob/34a921b9667402b6d90731f1fd9e3de1ef27f8bf/packages/aws-cdk/lib/api/cxapp/cloud-assembly.ts#L175-L216).
+We should in all likelihood re-use that mechanism.
+
+Note, however, that that requires access to the construct tree,
+which we don't necessarily have in a trivial way from all contexts
+(for example, for an access to a deprecated property of a struct interface).
+Which means we will need to add some mechanism to `core`
+that allows elements to register warnings with without necessarily having access to a construct,
+and which then will be gathered by synthesis and added to the resulting Cloud Assembly
+(parented below a certain construct, perhaps the `App`?).
+This also means we most likely won't be able to link the usage of the deprecated element to a particular `Stack`
+that it was used in.
+
+#### 2. Adding metadata when using deprecated elements
+
+The second element of this change is to actually add the appropriate metadata when any deprecated element is used.
+
+Before we dive into the possible solutions,
+I think it's worth it to talk about what solutions we _don't_ want,
+as that will strongly inform how we approach the problem.
+
+I think we can rule out any solution that requires parsing the customer's source code in search for usages of deprecated elements.
+The downsides of this approach include:
+
+* We would have to write a separate parser for each language supported by the CDK.
+* This additional parsing could have an adverse impact on the performance of CDK commands.
+
+If we agree that we don't want the above solution,
+then the only avenue we have left is injecting some extra code in the deprecated elements during compilation of the TypeScript code into JavaScript.
+There are two ways we can do that: through JSII, or purely through the TypeScript compiler.
+
+##### 2.1. Modifying JSII
+
+We could modify JSII to add a capability to it that allows injecting some code when compiling every deprecated element.
+It would involve writing a TypeScript transform,
+that adds some code whenever the deprecated element is accessed.
+Nick did a similar change in PR [#2348](https://github.com/aws/jsii/pull/2348)
+to add runtime version information to all JSII-compiled classes.
+
+Advantages:
+
+* Doing it in JSII wil make it easy to record the module the given element is in.
+
+Disadvantages:
```

Comment:
How would it work for struct properties?

Reply:
I guess for struct properties, we would have to detect their usage statically in JSII 😕.